### DISCUSSION TITLE
How to Crawl or Backfill data for a Search or Recommendation system?

### DISCUSSION THREAD

**Discussion by FaizShah** [bluesky-social/atproto#3054] [How to do it?] [answered]
Hi,

I was looking through the dev docs and it’s not clear to me what is the intended way for devs to acquire a training set to build Search or Recommendation applications.

What I see from the docs is you have the `backfill window` in the event stream but it explicitly says it’s not intended to be used to replay the history especially from the start: https://atproto.com/specs/event-stream

Then I see in the HTTP api there is a way to list all the repos [`com.atproto.sync.listRepos`](https://docs.bsky.app/docs/api/com-atproto-sync-list-repos) without auth but then you in order to [`com.atproto.repo.listRecords`](https://docs.bsky.app/docs/api/com-atproto-repo-list-records) in the repo it requires auth. Additionally this would be done as rate limited http calls instead of as a bulk operation like an event stream replay.

What is the intended way for developers to acquire a training set for Search or Recommendation applications?

**Comment by FaizShah** [comment, marked as answer]
Documentation here: https://docs.bsky.app/docs/advanced-guides/backfill

It seems that enumerating it using the API via listerepos getrepo is the official solution. Seems inefficient with no  batch op tho would prefer to replay the stream.

**Reply by DavidBuchanan314** [reply]
If you do this with enough concurrency then network bandwidth is the bottleneck, so I'm not sure I'd call it inefficient per se

**Comment by rozhok** [comment]
Is there any better ways to perform a full backfill?

**Reply by yamarten** [reply]
It is still under development, but it will be best to use [tap](https://github.com/bluesky-social/indigo/pull/1170) in the future.

At the protocol level, get a repository list from the relay and then download CAR from each PDS. [slinky](https://tangled.org/bnewbold.net/cobalt/tree/main/cmd/slinky) or [zeppelin](https://github.com/zeppelin-social/backfill-bsky) may be helpful.
